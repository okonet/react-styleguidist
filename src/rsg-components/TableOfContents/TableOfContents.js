import React, { Component, PropTypes } from 'react';
import { filterComponentsByName, getFilterRegExp } from '../../utils/utils';
import ComponentsList from 'rsg-components/ComponentsList';
import TableOfContentsRenderer from 'rsg-components/TableOfContents/TableOfContentsRenderer';

export default class TableOfContents extends Component {
	static propTypes = {
		sections: PropTypes.array.isRequired,
	};

	constructor(props) {
		super(props);

		this.state = {
			searchTerm: '',
		};
	}

	getComponents(components, searchTerm) {
		return filterComponentsByName(components || [], searchTerm);
	}

	getSections(sections = [], searchTerm) {
		const regExp = getFilterRegExp(searchTerm);
		return sections.reduce((filteredSections, { name, components: subComponents = [], sections: subSections }) => {
			subComponents = this.getComponents(subComponents, searchTerm);
			if (subComponents.length || !searchTerm || regExp.test(name)) {
				filteredSections.push({
					heading: true,
					name,
					content: this.renderLevel(subComponents, subSections, searchTerm),
				});
			}
			return filteredSections;
		}, []);
	}

	renderLevel(components, sections, searchTerm) {
		const items = [
			...this.getComponents(components, searchTerm),
			...this.getSections(sections, searchTerm),
		];
		return (
			<ComponentsList items={items} />
		);
	}

	render() {
		const { searchTerm } = this.state;
		const { sections } = this.props;

		// If there is only one section, we treat it as a root section
		// In this case the name of the section won't be rendered and it won't get left padding
		const content = sections.length === 1
			? this.renderLevel(sections[0].components, sections[0].sections, searchTerm)
			: this.renderLevel(null, sections, searchTerm);
		return (
			<TableOfContentsRenderer
				searchTerm={searchTerm}
				onSearchTermChange={searchTerm => this.setState({ searchTerm })}
			>
				{content}
			</TableOfContentsRenderer>
		);
	}
}
