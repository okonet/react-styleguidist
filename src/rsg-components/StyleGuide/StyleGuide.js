import React, { Component, PropTypes } from 'react';
import isEmpty from 'lodash/isEmpty';
import TableOfContents from 'rsg-components/TableOfContents';
import Message from 'rsg-components/Message';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import { DOCS_CONFIG, HOMEPAGE } from '../../../scripts/consts';

export default class StyleGuide extends Component {
	static propTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		sidebar: PropTypes.bool,
		singleExample: PropTypes.bool,
	};

	static childContextTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		singleExample: PropTypes.bool,
	};

	static defaultProps = {
		sidebar: true,
	};

	getChildContext() {
		return {
			codeKey: this.props.codeKey,
			config: this.props.config,
			singleExample: this.props.singleExample,
		};
	}

	render() {
		const { config, sections, sidebar } = this.props;

		return (
			<StyleGuideRenderer
				title={config.title}
				homepageUrl={HOMEPAGE}
				toc={<TableOfContents sections={sections} />}
				sidebar={isEmpty(sections) ? false : sidebar}
			>
				{
					!isEmpty(sections)
						? <Sections sections={sections} sidebar={sidebar} />
						: <Message>
								No components or sections found.
								Check [the `components` and `sections` options]({DOCS_CONFIG}) in your style guide
								config.
							</Message>
				}
			</StyleGuideRenderer>
		);
	}
}
