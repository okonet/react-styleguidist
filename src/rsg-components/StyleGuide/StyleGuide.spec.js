import React from 'react';
import TableOfContents from '../TableOfContents';
import StyleGuide from './StyleGuide';
import { StyleGuideRenderer } from './StyleGuideRenderer';

const components = [
	{
		name: 'Foo',
		pathLine: 'components/foo.js',
		filepath: 'components/foo.js',
		props: {
			description: 'Foo foo',
		},
	},
	{
		name: 'Bar',
		pathLine: 'components/bar.js',
		filepath: 'components/bar.js',
		props: {
			description: 'Bar bar',
		},
	},
];
const sections = [{ components }];
const config = {
	title: 'Hello',
};

it('should render components list', () => {
	const actual = shallow(
		<StyleGuide
			codeKey={1}
			config={config}
			sections={sections}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should render error message when there is no components and sections', () => {
	const actual = shallow(
		<StyleGuide
			codeKey={1}
			config={config}
			sections={[]}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('renderer should render logo, table of contents and passed children', () => {
	const actual = shallow(
		<StyleGuideRenderer
			classes={{}}
			title={config.title}
			toc={<TableOfContents components={components} sections={sections} />}
			homepageUrl="http://react-styleguidist.js.org/"
			sidebar
		>
			<h1>Content</h1>
		</StyleGuideRenderer>
	);

	expect(actual).toMatchSnapshot();
});
