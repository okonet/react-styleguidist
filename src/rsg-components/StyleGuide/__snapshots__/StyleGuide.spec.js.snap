exports[`test renderer should render logo, table of contents and passed children 1`] = `
<div
  className="">
  <main>
    <div>
      <h1>
        Content
      </h1>
      <footer>
        <_class
          text="Generated with [React Styleguidist](http://react-styleguidist.js.org/)" />
      </footer>
    </div>
  </main>
  <div>
    <div>
      <_class>
        Hello
      </_class>
    </div>
    <TableOfContents
      components={
        Array [
          Object {
            "filepath": "components/foo.js",
            "name": "Foo",
            "pathLine": "components/foo.js",
            "props": Object {
              "description": "Foo foo",
            },
          },
          Object {
            "filepath": "components/bar.js",
            "name": "Bar",
            "pathLine": "components/bar.js",
            "props": Object {
              "description": "Bar bar",
            },
          },
        ]
      }
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      } />
  </div>
</div>
`;

exports[`test should render components list 1`] = `
<_class
  homepageUrl="https://github.com/styleguidist/react-styleguidist"
  sidebar={true}
  title="Hello"
  toc={
    <TableOfContents
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      } />
  }>
  <Sections
    sections={
      Array [
        Object {
          "components": Array [
            Object {
              "filepath": "components/foo.js",
              "name": "Foo",
              "pathLine": "components/foo.js",
              "props": Object {
                "description": "Foo foo",
              },
            },
            Object {
              "filepath": "components/bar.js",
              "name": "Bar",
              "pathLine": "components/bar.js",
              "props": Object {
                "description": "Bar bar",
              },
            },
          ],
        },
      ]
    }
    sidebar={true} />
</_class>
`;

exports[`test should render error message when there is no components and sections 1`] = `
<_class
  homepageUrl="https://github.com/styleguidist/react-styleguidist"
  sidebar={false}
  title="Hello"
  toc={
    <TableOfContents
      sections={Array []} />
  }>
  <_class>
    No components or sections found. Check [the \`components\` and \`sections\` options](
    https://github.com/styleguidist/react-styleguidist/blob/master/docs/Configuration.md
    ) in your style guide config.
  </_class>
</_class>
`;
